text/0763-asset-importing-spec.md

```diff
@@ -0,0 +1,200 @@
+
+```js
+import myImage from './hello.png';
+class extends Component {
+  myImage = myImage
+}
+```
+```hbs
+<img src={{this.myImage}} />
+```
+
+## Motivation
+
+Apps and addons often need to refer to images, fonts, etc from within their JS, HBS,CSS, or HTML files. These references need to remain correct even as both the referring file and the referred-to file go through a build that might include fingerprinting and optimization.
+
+Today, the ecosystem mostly relies on [broccoli-asset-rev](https://github.com/ember-cli/broccoli-asset-rev). But broccoli-asset-rev has some major architectural problems and it doesn't take advantage of the newer capabilities we have in ember-auto-import and embroider.
+
+The biggest problem with broccoli-asset-rev is that it cannot achieve full correctness because it tries to detect inter-file references within arbitrary source code that is not statically analyzable. It's easy to accidentally refactor working code into code that works in development and test and then fails in production! For example, assuming broccoli-asset-rev is configured to fingerprint PNG assets, this refactor will look correct until you get into production and then have broken images:
+
+```diff
+class extends Component {
+-  iconURL = "/icons/happy.png"
++  get iconURL() {
++    return `/icons/${this.args.emotion}.png`
++  }
+}
+```
+
+The regular expressions that broccoli-asset-rev uses to do reference rewriting are effectively un-editable at this point. Any possible change you'd make to them breaks somebody's app.
+
+Another problem with broccoli-asset-rev is that it's "push based", meaning it handles files because they exist, rather than handling files because somebody is actually consuming them. It can't detect unused assets and it can't detect missing assets at build time. It exposes the underlying implementation of how assets get delivered, making it hard to change later without breaking app code.
+
+And because broccoli-asset-rev exposes assumptions about the exact layout of the final build output, it's not compatible with Embroider. Embroider allows different JavaScript packagers to optimize the app delivery in any spec-compliant way. For example, small images might get inlined as `data:` URIs, CSS `@import()` might get inlined or not, etc. All these concerns really need to be integrated into the JavaScript packager, because asset handling affects app-wide hashing & fingerprinting.
+
+In contrast, a pull-based design that lets code declare what assets it needs and then not worry about how those assets will get delivered is safer and easier to change in the future.
+
+## Detailed design
+
+Inter-file references take many forms. Here is a guide to inter-file references organized by which kind of file is doing the referring ("From Type") and which type of file it's referring to ("To Type"). Only the row containing **Proposed** is new in this RFC:
+
+From Type | To Type | Example Syntax | Semantics
```

Comment:
Today when folks use `.wasm` [one approach is](https://developer.mozilla.org/en-US/docs/WebAssembly/Loading_and_running) is to use `fetch` as follows:

```js
WebAssembly.instantiateStreaming(fetch('simple.wasm'), importObject)
.then(results => {
  // Do something with the results!
});
```

This also runs into the same finger printing + reachability problems described in this RFC.

My concern arises when attempting to apply this RFCs solution to the `wasm` problem. This RFC would recommend:

```js
import simple from './simple.wasm';

WebAssembly.instantiateStreaming(fetch(simpl), importObject)
.then(results => {
  // Do something with the results!
});
```

Which I suspect would work today, BUT it wouldn't surprise me if `import simple from './simple.wasm'` [eventually becomes an official way to load and import `wasm`](https://github.com/WebAssembly/esm-integration/tree/master/proposals/esm-integration), at which the RFC's approach would be incompatible....

Reply:
To expand this beyond `wasm`, I am not super comfortable with the idea of creating "meaning" to the default export of a currently unsupported module FileType. `wasm` above is an example that may very likely become supported, and I would be extremely surprised if we don't see further evolution in this area.

To mitigate said risk, we may want to consider adding this capability too `@embroider/macros`. Specifically, in additional to `importSync` `importURL` could be created.

```js
import { importURL } from '@embroider/macros';

WebAssembly.instantiateStreaming(importURL('simple.wasm'), importObject)
.then(results => {
  // Do something with the results!
});
```

or 

```js
import { importURL } from '@embroider/macros';

importURL('thing/icon.png');
importURL('thing/style.css');
importURL('.'); // either illegal & throws at build/lint or implies the current module MUST be addressable independently of a larger bundle
```

Reply:
Just to add some increased specificity, ultimately this RFC is about being able to gain access to the `import.meta.url` for a given `ModuleSpecifier` a form of module/bundler/runtime reflection but in a statically analyzable manner.

Given that, a more pedantic name could be:

```js
import { importMeta } from '@embroider/macros'
importMeta(ModuleSpecifier).url
```

Reply:
Ok some further snooping it looks like `node.js` has an experimental feature which I think is an even closer fit to what this RFC is exploring:

Specifically: [`import.meta.resolve(specifier[, parent])`](https://nodejs.org/api/esm.html#esm_import_meta_resolve_specifier_parent)

`import.meta.resolve` is async, so for our needs `import.meta.resolveSync` variant would be important.
One could implement `import.meta.resolveSync` via code-mode or more conservatively as an `@embroider/macros` macro:

```js
import { resolveSync } from '@embroider/macros';
```

```hbs
<img src={{resolve-sync 'thing.jpg'}} >
```



----

Although it seems fishy at first to introduce `import.meta.resolveSync` it is actually pretty aligned with the design of `import.meta` as it is designed to be augmented by a host, as one can see with Nodes `import.meta.resolve`.

Reply:
Lots of good thoughts here that I agree with as well.

Reply:
Yeah, this is pretty convincing, I think we can iterate in this direction.